File: stack/3430_hard/subarrays_max_min_sum.cpp
#include <deque>
#include <vector>
using namespace std;

long long computeMaxMinSum(vector<int> &nums, int k) //  LeetCode Q.3430.
{
    long long totalMaxMinSum = 0;
    long long windowMaxSum = 0, windowMinSum = 0;

    // Format: {idx, num, shares}. Use long long to prevent overflow.
    deque<tuple<int, int, long long>> maxStack, minStack;

    for (int endIdx = 0; endIdx < nums.size(); endIdx++)
    {
        int startIdx = max(0, endIdx - k + 1);

        // Window start idx slides by 1: must update stacks' info.
        if (startIdx > 0)
        {
            get<2>(maxStack.front())--; // Decrement stack's front num shares.
            windowMaxSum -= get<1>(maxStack.front());

            // Front num out of window.
            if (get<0>(maxStack.front()) < startIdx)
                maxStack.pop_front();

            get<2>(minStack.front())--; // Decrement stack's front num shares.
            windowMinSum -= get<1>(minStack.front());

            // Front num out of window.
            if (get<0>(minStack.front()) < startIdx)
                minStack.pop_front();
        }

        long long num = nums[endIdx];

        while (!maxStack.empty() && get<1>(maxStack.back()) <= num)
        {
            windowMaxSum -= get<1>(maxStack.back()) * get<2>(maxStack.back());
            maxStack.pop_back();
        }

        long long maxShares = min(endIdx + 1, k);
        if (!maxStack.empty())
            maxShares = endIdx - get<0>(maxStack.back());

        maxStack.push_back({endIdx, num, maxShares});
        windowMaxSum += num * maxShares;

        while (!minStack.empty() && get<1>(minStack.back()) >= num)
        {
            windowMinSum -= get<1>(minStack.back()) * get<2>(minStack.back());
            minStack.pop_back();
        }

        long long minShares = min(endIdx + 1, k);
        if (!minStack.empty())
            minShares = endIdx - get<0>(minStack.back());

        minStack.push_back({endIdx, num, minShares});
        windowMinSum += num * minShares;

        totalMaxMinSum += windowMaxSum + windowMinSum;
    }

    return totalMaxMinSum;
}

File: docs/stack/3430_hard/subarrays_max_min_sum.py

def compute_max_min_sum(nums: list[int], k: int) -> int:  # LeetCode Q.3430.
    subarrays_max_min_sum = 0

    max_stack: list[list[int]] = []  # Format: [idx, num, shares].
    subarrays_max_sum = 0

    min_stack: list[list[int]] = []  # Format: [idx, num, shares].
    subarrays_min_sum = 0

    for end_idx, num in enumerate(nums):
        start_idx = max(0, end_idx - k + 1)

        # Window start idx slides by 1: must update stacks' info.
        if start_idx > 0:
            max_stack[0][2] -= 1  # Decrement stack's front num shares.
            subarrays_max_sum -= max_stack[0][1]

            if max_stack[0][0] < start_idx:  # Front num out of window.
                max_stack.pop(0)

            min_stack[0][2] -= 1  # Decrement stack's front num shares.
            subarrays_min_sum -= min_stack[0][1]

            if min_stack[0][0] < start_idx:  # Front num out of window.
                min_stack.pop(0)

        while max_stack and max_stack[-1][1] <= num:
            _, prev_num, prev_shares = max_stack.pop(-1)
            subarrays_max_sum -= prev_num * prev_shares

        if not max_stack:
            shares = min(end_idx + 1, k)

        else:
            shares = end_idx - max_stack[-1][0]

        subarrays_max_sum += num * shares
        max_stack.append([end_idx, num, shares])

        while min_stack and min_stack[-1][1] >= num:
            _, prev_num, prev_shares = min_stack.pop(-1)
            subarrays_min_sum -= prev_num * prev_shares

        if not min_stack:
            shares = min(end_idx + 1, k)

        else:
            shares = end_idx - min_stack[-1][0]

        subarrays_min_sum += num * shares
        min_stack.append([end_idx, num, shares])

        subarrays_max_min_sum += subarrays_max_sum + subarrays_min_sum

    return subarrays_max_min_sum
